docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,546 @@
+accordingly (probably by erroring out).  Additionally, a released array
+MAY set :c:member:`ArrowArray.release` to NULL.
+
+Release callback semantics -- for consumers
+'''''''''''''''''''''''''''''''''''''''''''
+
+Consumers MUST call an array's release callback when they won't be using
+it anymore, but they MUST not call any of its child arrays' release callbacks
+(including the optional dictionary).  The producer is responsible for releasing
+the children.
+
+Consumers MUST check that the release callback is non-NULL before calling it.
+The release callback being NULL is not an error, it should just be ignored.
+
+In any case, a consumer MUST not try to access the ArrowArray struct anymore
+after calling its release callback -- including any associated data such
+as its children.
+
+Release callback semantics -- for producers
+'''''''''''''''''''''''''''''''''''''''''''
+
+If producers need additional information for lifetime handling (for
+example, a C++ producer may want to use ``shared_ptr`` for array and
+buffer lifetime), they MUST use the :c:member:`ArrowArray.private_data`
+member to locate the required bookkeeping information.
+
+The release callback MUST not assume that the struct will be located
+at the same memory location as when it was originally produced.  The consumer
+is free to move the struct around (see "Movability").
+
+The release callback MUST walk all children arrays (including the optional
+dictionary) and call their own release callbacks.
+
+The release callback MUST free any data area directly owned by the struct
+(such as the buffers and children arrays).
+
+The release callback MUST mark the array as released, by setting
+:c:member:`ArrowArray.format` to NULL.
+
+Additionally, the release callback MUST be idempotent, which is commonly
+achieved by setting itself to NULL.
+
+Movability
+''''''''''
+
+The consumer can *move* the ``ArrowArray`` struct by bitwise copying (or
+shallow member-wise copying).  Then it MUST mark the source struct released
+(see "released array" above for how to do it) but **without** calling the
+release callback.  This ensures that only one live copy of the struct is
+active at any given time and that lifetime is correctly communicated to
+the producer.
+
+Only the parent array (such as passed between producer and consumer) may be
+moved.  It is not recommended to move the children arrays.  It should also
+not be necessary, since the :c:member:`ArrowArray.children` and
+:c:member:`ArrowArray.dictionary` are typically heap-allocated by the
+producer.
+
+.. note::
+
+   For bitwise copying to work correctly, the pointers inside the struct
+   (including private_data) MUST not point inside the struct itself.
+   Also, external pointers to the struct MUST not be stored by the producer.
+   Instead, the producer MUST use the :c:member:`ArrowArray.private_data`
+   member so as to remember any necessary bookkeeping information.
+
+Example use case
```

Comment:
(Surely out of scope for this protocol draft but) it would be great to have an integration test demonstrating this use case, a blog post with code illustrating the setup, or both.

Reply:
Yeah, before adopting this spec, I should implement the C++ adaptation layer and if possible a trivial pure Python example.